json.lua

```diff
@@ -57,44 +57,56 @@ end
 
 
 local function encode_table(val, stack)
-  local res = {}
-  stack = stack or {}
-
-  -- Circular reference?
-  if stack[val] then error("circular reference") end
-
-  stack[val] = true
-
-  if rawget(val, 1) ~= nil or next(val) == nil then
-    -- Treat as array -- check keys are valid and it is not sparse
-    local n = 0
-    for k in pairs(val) do
-      if type(k) ~= "number" then
-        error("invalid table: mixed or invalid key types")
-      end
-      n = n + 1
-    end
-    if n ~= #val then
-      error("invalid table: sparse array")
-    end
-    -- Encode
-    for i, v in ipairs(val) do
-      table.insert(res, encode(v, stack))
-    end
-    stack[val] = nil
-    return "[" .. table.concat(res, ",") .. "]"
-
-  else
-    -- Treat as an object
-    for k, v in pairs(val) do
-      if type(k) ~= "string" then
-        error("invalid table: mixed or invalid key types")
-      end
-      table.insert(res, encode(k, stack) .. ":" .. encode(v, stack))
-    end
-    stack[val] = nil
-    return "{" .. table.concat(res, ",") .. "}"
-  end
+	local res = {}
+	stack = stack or {}
+
+	-- Circular reference?
+	if stack[val] then error("circular reference") end
+
+	stack[val] = true
+	-- Check whether to treat as a array or object
+	local array = true
+	local length = 0
+	local nLen = 0
+	local count = 0
+	for k,v in pairs(val) do
+		if (type(k) ~= "number" or k<=0) and not (k == "n" and type(v) == "number") then
```

Comment:
What about fractional keys, like `4.2`? I think you also need a check like `k % 1 == 0` or `math.floor(k) == k` to check that `k` is (not) an integer.

Reply:
Thank you, added k%1 ~= 0 in the condition.